test/drbgtest.c

```diff
@@ -251,30 +266,273 @@ static int test_drbg_reseed(int expect_success,
 
 
 #if defined(OPENSSL_SYS_UNIX)
+/* number of children to fork */
+#define DRBG_FORK_COUNT 9
+/* two results per child, two for the parent */
+#define DRBG_FORK_RESULT_COUNT (2 * (DRBG_FORK_COUNT + 1))
+
+typedef struct drbg_fork_result_st {
+
+    unsigned char random[RANDOM_SIZE]; /* random output */
+
+    int pindex;               /* process index (0: parent, 1,2,3...: children)*/
+    pid_t pid;                /* process id */
+    int private;              /* true if the private drbg was used */
+    char name[10];            /* 'parent' resp. 'child 1', 'child 2', ... */
+} drbg_fork_result;
+
+/*
+ * Sort the drbg_fork_result entries in lexicographical order
+ *
+ * This simplifies finding duplicate random output and makes
+ * the printout in case of an error more readable.
+ */
+static int compare_drbg_fork_result(const void * left, const void * right)
+{
+    int result;
+
+    const drbg_fork_result *l = left;
+    const drbg_fork_result *r = right;
+
+    /* separate public and private results */
+    result = l->private - r->private;
+
+    if (result == 0)
+        result = memcmp(l->random, r->random, RANDOM_SIZE);
+
+    if (result == 0)
+        result = l->pindex - r->pindex;
+
+    return result;
+}
+
+/*
+ * Sort two-byte chunks of random data
+ *
+ * Used for finding collisions in two-byte chunks
+ */
+static int compare_rand_chunk(const void * left, const void * right)
+{
+    return memcmp(left, right, 2);
+}
+
 /*
- * Test whether primary, public and private DRBG are reseeded after
- * forking the process.
+ * Test whether primary, public and private DRBG are reseeded
+ * in the child after forking the process. Collect the random
+ * output of the public and private DRBG and send it back to
+ * the parent process.
  */
-static int test_drbg_reseed_after_fork(EVP_RAND_CTX *primary,
-                                       EVP_RAND_CTX *public,
-                                       EVP_RAND_CTX *private)
+static int test_drbg_reseed_in_child(EVP_RAND_CTX *primary,
+                                     EVP_RAND_CTX *public,
+                                     EVP_RAND_CTX *private,
+                                     drbg_fork_result result[2])
 {
+    int rv = 0, status;
+    int fd[2];
     pid_t pid;
-    int status=0;
+    unsigned char random[2 * RANDOM_SIZE];
 
-    pid = fork();
-    if (!TEST_int_ge(pid, 0))
+    if (!TEST_int_ge(pipe(fd), 0))
```

Comment:
Is this mucking around with pipes necessary?  Couldn't the child processes write into separate buffers coordinated by the parent?

Reply:
Wouldn’t that require using shared memory to pass the results to the parent, and wouldn’t that be less portable than unnamed pipes, the most unixish way of interprocess communication?

Reply:
Ah, fork() not threads -- ignore this.  Shared memory is supported way back but pipes are older.